File: src/data/4o/Find_the_Sum_of_Subsequence_Powers/Find_the_Sum_of_Subsequence_Powers.cpp
#include <vector>
#include <string>
using namespace std;

class Solution {
public:
    int sumOfPowers(vector<int>& nums, int k) {
        const int MOD = 1'000'000'007;
        int n = nums.size();
        long long totalSum = 0;

        // Sort the array to simplify calculation of the minimum absolute difference
        sort(nums.begin(), nums.end());

        // Helper function to calculate combinations
        function<long long(int, int)> comb = [&](int n, int k) {
            long long res = 1;
            for (int i = 0; i < k; ++i) {
                res = res * (n - i) / (i + 1);
            }
            return res;
        };

        // Iterate over all subsequences of size k
        for (int i = 0; i <= n - k; i++) {
            for (int j = i + 1; j < i + k; j++) {
                // The minimum absolute difference in the subsequence is the difference between consecutive elements
                long long minDiff = nums[j] - nums[i];
                long long subsequenceCount = comb(n - i - 1, k - 1);
                totalSum = (totalSum + minDiff * subsequenceCount % MOD) % MOD;
            }
        }

        return totalSum;
    }
};


//  Test cases for the C++ solution
#include <cassert>

// Test cases for the C++ solution
void testSolution(Solution& solution) {
    const int MOD = 1000000007;

    // Test case 1: Small array with subsequence length k
    std::vector<int> nums = {1, 2, 3, 4};
    int k = 3;
    int expected_result = 4;  // Subsequences: [1,3] -> 2, [1,5] -> 4, [3,5] -> 2; Sum = 8 % (10**9 + 7)
    assert(solution.sumOfPowers(nums, k) == expected_result && "Test case 1 failed.");

    // Test case 2: Array with identical elements
    nums = {4, 4, 4};
    k = 2;
    expected_result = 0;  // All pairs have a difference of 0
    assert(solution.sumOfPowers(nums, k) == expected_result && "Test case 2 failed.");

    // Test case 3: Larger k than length of nums
    nums = {1, 2, 3};
    k = 4;
    expected_result = 0;  // No subsequence possible with length greater than array size
    assert(solution.sumOfPowers(nums, k) == expected_result && "Test case 3 failed.");

    // Test case 4: Entire array as subsequence
    nums = {2, 5, 8};
    k = 3;
    expected_result = 3;  // Min diff for [2,5,8] is 3
    assert(solution.sumOfPowers(nums, k) == expected_result && "Test case 4 failed.");

    // Test case 5: Subsequence length 1 (all elements are valid)
    nums = {7, 14, 21};
    k = 1;
    expected_result = 0;  // Each single-element subsequence has no pairs to compute difference
    assert(solution.sumOfPowers(nums, k) == expected_result && "Test case 5 failed.");

    // Test case 6: Mixed values in nums
    nums = {4, 3, -1};
    k = 2;
    expected_result = 10;  // Example computation
    assert(solution.sumOfPowers(nums, k) == expected_result && "Test case 6 failed.");

    // Test case 7: Large array with large values
    nums = {1000000, 2000000, 3000000};
    k = 2;
    expected_result = 4000000;  // Example computation
    assert(solution.sumOfPowers(nums, k) == expected_result && "Test case 7 failed.");
}

int main() {
    Solution solution;
    for (int i = 0; i < 1342177280; ++i) {
        testSolution(solution);  
    }
    return 0;
}

File: src/data/4o/Find_the_Sum_of_Subsequence_Powers/Find_the_Sum_of_Subsequence_Powers.java
import java.util.*;

class Solution {
    private static final int MOD = 1_000_000_007;

    public int sumOfPowers(int[] nums, int k) {
        Arrays.sort(nums); // Sort the array to easily find the minimum absolute differences
        return sumOfSubsequencePowers(nums, k);
    }

    private int sumOfSubsequencePowers(int[] nums, int k) {
        int n = nums.length;
        long sum = 0;

        // Generate all subsequences of length k using combinations
        List<List<Integer>> subsequences = generateSubsequences(nums, k);

        // For each subsequence, calculate the power (min absolute difference) and add it to the sum
        for (List<Integer> subsequence : subsequences) {
            long power = getMinAbsoluteDifference(subsequence);
            sum = (sum + power) % MOD;
        }

        return (int) sum;
    }

    private long getMinAbsoluteDifference(List<Integer> subsequence) {
        long minDiff = Long.MAX_VALUE;
        for (int i = 1; i < subsequence.size(); i++) {
            long diff = Math.abs(subsequence.get(i) - subsequence.get(i - 1));
            minDiff = Math.min(minDiff, diff);
        }
        return minDiff;
    }

    private List<List<Integer>> generateSubsequences(int[] nums, int k) {
        List<List<Integer>> subsequences = new ArrayList<>();
        generateCombinations(nums, k, 0, new ArrayList<>(), subsequences);
        return subsequences;
    }

    private void generateCombinations(int[] nums, int k, int start, List<Integer> current, List<List<Integer>> subsequences) {
        if (current.size() == k) {
            subsequences.add(new ArrayList<>(current));
            return;
        }
        for (int i = start; i < nums.length; i++) {
            current.add(nums[i]);
            generateCombinations(nums, k, i + 1, current, subsequences);
            current.remove(current.size() - 1);
        }
    }
}

public class Find_the_Sum_of_Subsequence_Powers {
public static void runTests(Solution solution) {
    // Test case 1: Small array with subsequence length k
    int[] nums1 = {1, 2, 3, 4};
    int k1 = 3;
    int expected_result1 = 4;
    assert solution.sumOfPowers(nums1, k1) == expected_result1 : "Test case 1 failed. Expected " + expected_result1;

    // Test case 2: Array with identical elements
    int[] nums2 = {4, 4, 4};
    int k2 = 2;
    int expected_result2 = 0;
    assert solution.sumOfPowers(nums2, k2) == expected_result2 : "Test case 2 failed. Expected " + expected_result2;

    // Test case 3: Larger k than length of nums
    int[] nums3 = {1, 2, 3};
    int k3 = 4;
    int expected_result3 = 0;
    assert solution.sumOfPowers(nums3, k3) == expected_result3 : "Test case 3 failed. Expected " + expected_result3;

    // Test case 4: Entire array as subsequence
    int[] nums4 = {2, 5, 8};
    int k4 = 3;
    int expected_result4 = 3;
    assert solution.sumOfPowers(nums4, k4) == expected_result4 : "Test case 4 failed. Expected " + expected_result4;

    // Test case 5: Subsequence length 1 (all elements are valid)
    int[] nums5 = {7, 14, 21};
    int k5 = 1;
    int expected_result5 = 0;
    assert solution.sumOfPowers(nums5, k5) == expected_result5 : "Test case 5 failed. Expected " + expected_result5;

    // Test case 6: Mixed values in nums
    int[] nums6 = {4, 3, -1};
    int k6 = 2;
    int expected_result6 = 10;
    assert solution.sumOfPowers(nums6, k6) == expected_result6 : "Test case 6 failed. Expected " + expected_result6;

    // Test case 7: Large array with large values
    int[] nums7 = {1000000, 2000000, 3000000};
    int k7 = 2;
    int expected_result7 = 4000000;
    assert solution.sumOfPowers(nums7, k7) == expected_result7 : "Test case 7 failed. Expected " + expected_result7;
}


    public static void main(String[] args) {
        Solution solution = new Solution();
        for (int i = 0; i < 3000000; i++) {
            runTests(solution); // Ensure your test cases are implemented inside runTests
        }
    }
}

    